File: backend/src/main/java/at/ac/tuwien/sepr/groupphase/backend/service/EventService.java
package at.ac.tuwien.sepr.groupphase.backend.service;

import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.event.DetailedEventDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.event.EventCreateDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.event.EventUpdateDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.event.SimpleEventDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.event.TopTenEventDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.event.SeatmapDto;
import at.ac.tuwien.sepr.groupphase.backend.entity.Event;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import java.time.LocalDateTime;
import java.util.List;

/**
 * Service for managing events.
 */
public interface EventService {

    /**
     * Find all events ordered by date (ascending).
     *
     * @return list of all events
     */
    List<Event> findAll();

    /**
     * Find event by ID.
     *
     * @param id the event ID
     * @return the event entity
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException if event not found
     */
    Event findById(Long id);

    /**
     * Create a new event.
     * Validates event data before creation.
     *
     * @param event the event to create
     * @return the created event
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.ValidationException if event data is invalid
     */
    Event create(Event event);

    /**
     * Update an existing event.
     * Validates event data before update.
     *
     * @param event the event to update
     * @return the updated event
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException   if event not found
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.ValidationException if event data is invalid
     */
    Event update(Event event);

    /**
     * Delete an event by ID.
     * Cannot delete events with existing tickets.
     *
     * @param id the event ID
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException   if event not found
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.ValidationException if event has tickets
     */
    void delete(Long id);

    /**
     * Search events by title (case-insensitive, partial match).
     *
     * @param title the title to search for
     * @return list of matching events
     */
    List<Event> searchByTitle(String title);

    /**
     * Search events by type (case-insensitive, exact match).
     *
     * @param type the event type
     * @return list of matching events
     */
    List<Event> searchByType(String type);

    /**
     * Search events by duration with tolerance.
     * Finds events within duration ± tolerance minutes.
     *
     * @param duration  the target duration in minutes
     * @param tolerance the tolerance range in minutes
     * @return list of matching events
     */
    List<Event> searchByDuration(int duration, int tolerance);

    /**
     * Search events by description (case-insensitive, partial match).
     *
     * @param description the description text to search for
     * @return list of matching events
     */
    List<Event> searchByDescription(String description);

    /**
     * Search events by a general search term.
     * Searches across title, description, and other fields.
     *
     * @param searchTerm the search term
     * @return list of matching events
     */
    List<Event> searchEvents(String searchTerm);

    /**
     * Find all events for a specific artist.
     *
     * @param artistId the artist ID
     * @return list of events featuring this artist
     */
    List<Event> findByArtistId(Long artistId);

    /**
     * Find all events at a specific location.
     *
     * @param locationId the location ID
     * @return list of events at this location
     */
    List<Event> findByLocationId(Long locationId);

    /**
     * Search events within a date range.
     *
     * @param start the start date/time
     * @param end   the end date/time
     * @return list of events within the range
     */
    List<Event> searchByDateRange(LocalDateTime start, LocalDateTime end);

    /**
     * Search events by multiple criteria (all optional).
     * Combines all provided criteria with AND logic.
     *
     * @param title       the title to search for (optional)
     * @param type        the event type (optional)
     * @param minDuration minimum duration in minutes (optional)
     * @param maxDuration maximum duration in minutes (optional)
     * @param startDate   earliest event date/time (optional)
     * @param endDate     latest event date/time (optional)
     * @param locationId  the location ID (optional)
     * @return list of matching events
     */
    List<Event> searchByCriteria(
        String title,
        String type,
        Integer minDuration,
        Integer maxDuration,
        LocalDateTime startDate,
        LocalDateTime endDate,
        Long locationId
    );

    /**
     * Search events by time range and price range.
     * Finds events with at least one ticket in the price range.
     *
     * @param startDate  earliest event date/time (optional)
     * @param endDate    latest event date/time (optional)
     * @param minPrice   minimum ticket price in cents (optional)
     * @param maxPrice   maximum ticket price in cents (optional)
     * @param locationId the location ID (optional)
     * @return list of matching events
     */
    List<Event> searchByTimeAndPrice(
        LocalDateTime startDate,
        LocalDateTime endDate,
        Integer minPrice,
        Integer maxPrice,
        Long locationId
    );

    /**
     * Validate event data.
     * Checks date is in future, location exists, and title is present.
     *
     * @param event the event to validate
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.ValidationException if validation fails
     */
    void validateEventData(Event event);

    /**
     * Find top 10 events by ticket sales for a specific month and year.
     * Optionally filter by event type.
     *
     * @param month the month (1-12)
     * @param year  the year
     * @param type  the event type filter (optional)
     * @return list of top 10 events by ticket count
     */
    List<TopTenEventDto> findTopTenByTicketSales(int month, int year, String type);

    /**
     * Get seatmap for an event including seat status (free/reserved/sold).
     *
     * @param eventId the event ID
     * @return the seatmap with all seats and their status
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException if event not found
     */
    SeatmapDto getSeatmap(Long eventId);

    /**
     * Find minimum ticket price for an event.
     * Returns null if no tickets/prices exist.
     *
     * @param eventId the event ID
     * @return the minimum price in cents, or null if no price found
     */
    Integer findMinPriceForEvent(Long eventId);

    /**
     * Upload an image for an event.
     * Validates file size (max 5MB) and type (JPEG/PNG/WebP).
     *
     * @param eventId the event ID
     * @param image   the image file
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException   if event not found
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.ValidationException if image is invalid
     */
    void uploadImage(Long eventId, MultipartFile image);

    /**
     * Get image data for an event.
     *
     * @param eventId the event ID
     * @return the image bytes
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException if event or image not found
     */
    byte[] getImage(Long eventId);

    /**
     * Get image content type for an event.
     *
     * @param eventId the event ID
     * @return the MIME type (e.g., "image/jpeg")
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException if event or image not found
     */
    String getImageContentType(Long eventId);

    /**
     * Find all events ordered by date with pagination.
     *
     * @param pageable pagination information
     * @return page of events as DTOs
     */
    Page<SimpleEventDto> findAllAsDto(Pageable pageable);

    /**
     * Get event by ID with enriched price information as detailed DTO.
     * Includes location, artists, and minimum ticket price.
     *
     * @param id the event ID
     * @return the event as detailed DTO
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException if event not found
     */
    DetailedEventDto findByIdAsDto(Long id);

    /**
     * Create event from DTO including location and artists resolution.
     * Resolves location and artist IDs to actual entities before creation.
     *
     * @param dto the event creation data
     * @return the created event as DTO with enriched price
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.ValidationException if data is invalid
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException   if location or artist not found
     */
    SimpleEventDto createFromDto(EventCreateDto dto);

    /**
     * Update event from DTO including location and artists resolution.
     * Resolves location and artist IDs to actual entities before update.
     *
     * @param id  the event ID (must match DTO ID)
     * @param dto the event update data
     * @return the updated event as detailed DTO
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.ValidationException if data is invalid or IDs don't match
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException   if event, location, or artist not found
     */
    DetailedEventDto updateFromDto(Long id, EventUpdateDto dto);

    /**
     * Advanced event search with pagination.
     *
     * @param title      event title (optional)
     * @param type       event type (optional)
     * @param duration   target duration with ±30min tolerance (optional)
     * @param dateFrom   earliest event date (optional)
     * @param dateTo     latest event date (optional)
     * @param locationId location ID (optional)
     * @param priceMin   minimum ticket price in cents (optional)
     * @param priceMax   maximum ticket price in cents (optional)
     * @param pageable   pagination information
     * @return page of matching events
     */
    Page<SimpleEventDto> searchEventsAsDto(
        String title,
        String type,
        Integer duration,
        LocalDateTime dateFrom,
        LocalDateTime dateTo,
        Long locationId,
        Integer priceMin,
        Integer priceMax,
        Pageable pageable
    );

    /**
     * Find events by location ID and return as DTOs with enriched prices.
     *
     * @param locationId the location ID
     * @return list of events at this location as DTOs
     */
    List<SimpleEventDto> findByLocationIdAsDto(Long locationId);

    /**
     * Find events by artist ID and return as DTOs with enriched prices.
     *
     * @param artistId the artist ID
     * @return list of events featuring this artist as DTOs
     */
    List<SimpleEventDto> findByArtistIdAsDto(Long artistId);

    /**
     * Delete the image of an event.
     *
     * @param eventId the event ID
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException if event not found
     */
    void deleteImage(Long eventId);
}

File: frontend/src/app/services/event.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {catchError, map, Observable, of} from 'rxjs';
import {Globals} from '../global/globals';
import {Event} from '../dtos/event';
import {EventDetail} from '../dtos/event-detail';
import {TopTenEvent} from '../dtos/top-ten-event';
import {Page} from '../dtos/page';

/**
 * Service for managing event-related operations.
 * Provides functionality for event creation, search, updates, and retrieval of event data.
 * Handles communication with the backend events API endpoint.
 */
@Injectable({
  providedIn: 'root'
})
export class EventService {
  private baseUri: string = this.globals.backendUri + '/events';
  private imageLoadTimestamp: number = new Date().getTime();

  constructor(
    private httpClient: HttpClient,
    private globals: Globals
  ) {
  }

  /**
   * Retrieves all events from the system with pagination.
   *
   * @param page - Page number (0-indexed)
   * @param size - Number of events per page
   * @returns Observable of paginated events
   */
  getAllEvents(page: number = 0, size: number = 12): Observable<Page<Event>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    return this.httpClient.get<Page<Event>>(this.baseUri, {params});
  }

  /**
   * Searches for events using multiple filter criteria with pagination.
   * All parameters are optional and can be combined for advanced filtering.
   *
   * @param title - Optional title search term (supports partial matching)
   * @param type - Optional event type filter (e.g., 'concert', 'theater', 'opera', 'festival')
   * @param duration - Optional duration in minutes (applies ±30 minute tolerance)
   * @param dateFrom - Optional start date filter in ISO 8601 format (e.g., '2026-01-19T12:00:00')
   * @param dateTo - Optional end date filter in ISO 8601 format
   * @param locationId - Optional location ID to filter events by venue
   * @param priceMin - Optional minimum price in cents (e.g., 1000 for €10.00)
   * @param priceMax - Optional maximum price in cents
   * @param page - Page number (0-indexed)
   * @param size - Number of events per page
   * @returns Observable of paginated events matching the search criteria
   */
  searchEvents(
    title?: string,
    type?: string,
    duration?: number,
    dateFrom?: string,
    dateTo?: string,
    locationId?: number,
    priceMin?: number,
    priceMax?: number,
    page: number = 0,
    size: number = 12
  ): Observable<Page<Event>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    if (title) {
      params = params.set('title', title);
    }

    if (type) {
      params = params.set('type', type);
    }

    if (duration) {
      params = params.set('duration', duration.toString());
    }

    if (dateFrom) {
      params = params.set('dateFrom', dateFrom);
    }

    if (dateTo) {
      params = params.set('dateTo', dateTo);
    }

    if (locationId) {
      params = params.set('locationId', locationId.toString());
    }

    if (priceMin) {
      params = params.set('priceMin', priceMin.toString());
    }

    if (priceMax) {
      params = params.set('priceMax', priceMax.toString());
    }

    return this.httpClient.get<Page<Event>>(`${this.baseUri}/search`, {params});
  }

  /**
   * Retrieves detailed information for a specific event by its ID.
   *
   * @param id - The unique identifier of the event
   * @returns Observable of the event details
   * @throws NotFoundException if the event with the given ID does not exist
   */
  getEventById(id: number): Observable<EventDetail> {
    return this.httpClient.get<EventDetail>(`${this.baseUri}/${id}`);
  }

  /**
   * Retrieves the top ten events by ticket sales for a specific month and optional category.
   * Used for displaying popular events and sales statistics.
   *
   * @param month - The month number (1-12) for which to retrieve top events
   * @param year - The year for which to retrieve top events
   * @param type - Optional event type to filter by category (e.g., 'concert', 'theater')
   * @returns Observable of the top ten events with sales information
   */
  getTopTen(month: number, year: number, type?: string): Observable<TopTenEvent[]> {
    let params = new HttpParams()
      .set('month', month.toString())
      .set('year', year.toString());

    if (type) {
      params = params.set('type', type);
    }

    return this.httpClient.get<TopTenEvent[]>(`${this.baseUri}/top-ten`, {params});
  }

  /**
   * Creates a new event in the system.
   * Requires administrator privileges.
   *
   * @param event - The event data to create
   * @param event.title - The title/name of the event
   * @param event.type - The event type/category
   * @param event.durationMinutes - Optional duration in minutes
   * @param event.description - Optional detailed description
   * @param event.dateTime - The date and time of the event
   * @param event.locationId - The ID of the location/venue where the event takes place
   * @param event.artistIds - Optional array of artist IDs performing at the event
   * @returns Observable of the created event with assigned ID
   * @throws ValidationException if the event data is invalid
   * @throws NotFoundException if the referenced location or artists do not exist
   */
  createEvent(event: {
    title: string;
    type: string;
    durationMinutes?: number;
    description?: string;
    dateTime: string;
    locationId: number;
    artistIds?: number[];
  }): Observable<Event> {
    return this.httpClient.post<Event>(this.baseUri, event);
  }

  /**
   * Updates an existing event with new data.
   * Requires administrator privileges.
   *
   * @param id - The unique identifier of the event to update
   * @param event - The updated event data
   * @param event.id - Must match the id parameter
   * @param event.title - The updated title/name of the event
   * @param event.type - The updated event type/category
   * @param event.durationMinutes - Optional updated duration in minutes
   * @param event.description - Optional updated description
   * @param event.dateTime - The updated date and time
   * @param event.locationId - The updated location/venue ID
   * @param event.artistIds - Optional updated array of artist IDs
   * @returns Observable of the updated event
   * @throws ValidationException if the event data is invalid
   * @throws NotFoundException if the event, location, or artists do not exist
   * @throws ConflictException if tickets have already been sold and the update would conflict
   */
  updateEvent(id: number, event: {
    id: number;
    title: string;
    type: string;
    durationMinutes?: number;
    description?: string;
    dateTime: string;
    locationId: number;
    artistIds?: number[];
  }): Observable<EventDetail> {
    return this.httpClient.put<EventDetail>(`${this.baseUri}/${id}`, event);
  }

  /**
   * Deletes an event from the system.
   * Requires administrator privileges.
   *
   * @param id - The unique identifier of the event to delete
   * @returns Observable that completes when the event is deleted
   * @throws NotFoundException if the event does not exist
   * @throws ConflictException if tickets have already been sold for this event
   */
  deleteEvent(id: number): Observable<void> {
    return this.httpClient.delete<void>(`${this.baseUri}/${id}`);
  }

  /**
   * Uploads an image for a specific event.
   * The image will be used as the event's promotional/display image.
   *
   * @param eventId - The ID of the event to upload the image for
   * @param image - The image file to upload (supported formats: JPG, PNG)
   * @returns Observable that completes when the upload is successful
   * @throws ValidationException if the file format is not supported
   * @throws NotFoundException if the event does not exist
   */
  uploadImage(eventId: number, image: File): Observable<void> {
    const formData = new FormData();
    formData.append('image', image);
    return this.httpClient.post<void>(`${this.baseUri}/${eventId}/image`, formData);
  }

  /**
   * Deletes the image of a specific event.
   * Requires administrator privileges.
   *
   * @param eventId - The ID of the event whose image should be deleted
   * @returns Observable that completes when the image is deleted
   * @throws NotFoundException if the event does not exist or has no image
   */
  deleteImage(eventId: number): Observable<void> {
    return this.httpClient.delete<void>(`${this.baseUri}/${eventId}/image`);
  }

  /**
   * Constructs the URL for retrieving an event's image with cache busting.
   * Can be used directly in image src attributes.
   *
   * @param eventId - The ID of the event
   * @param bustCache - Optional: Force a new timestamp (default: false)
   * @returns The full URL to the event's image resource
   */
  getImageUrl(eventId: number, bustCache: boolean = false): string {
    const timestamp = bustCache ? new Date().getTime() : this.imageLoadTimestamp;
    return `${this.baseUri}/${eventId}/image?t=${timestamp}`;
  }

  /**
   * Invalidates the image cache for a specific event.
   * Call this after uploading a new image to force browser reload.
   */
  refreshImageCache(): void {
    this.imageLoadTimestamp = new Date().getTime();
  }

  /**
   * Checks if an event has an associated image without loading the full image data.
   *
   * @param eventId - The ID of the event to check
   * @returns Observable<boolean> - true if the event has an image, false otherwise
   */
  checkImageExists(eventId: number): Observable<boolean> {
    return this.httpClient.get(`${this.baseUri}/${eventId}/image`, {
      observe: 'response',
      responseType: 'blob'
    }).pipe(
      map((response) => response.status === 200),
      catchError(() => of(false))
    );
  }
}
